pallets/ajuna-awesome-avatars/src/lib.rs

```diff
@@ -40,15 +94,155 @@ pub mod pallet {
+				return Err(Error::<T>::SeasonStartsTooLate.into())
+			}
+
+			let season_id = Self::next_season_id();
+
+			if season_id > 0 {
+				if let Some(season) = Seasons::<T>::get(season_id - 1) {
+					if Season::are_seasons_overlapped(&season, &new_season) {
+						return Err(Error::<T>::EarlyAccessStartsTooEarly.into())
+					}
+				}
+			}
+
+			// save season
+			Seasons::<T>::insert(season_id, new_season.clone());
+
+			// increase next season id
+			match season_id.checked_add(1) {
+				Some(number) => NextSeasonId::<T>::put(number),
+				None => return Err(ArithmeticError::Overflow.into()),
+			};
+
+			Self::deposit_event(Event::NewSeasonCreated(new_season));
+
+			Ok(())
+		}
+
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+		pub fn update_season(
+			origin: OriginFor<T>,
+			season_id: SeasonId,
+			season: Season<BlockNumberFor<T>>,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			if season.is_early_access_start_too_late() {
+				return Err(Error::<T>::EarlyAccessStartsTooLate.into())
+			}
+
+			if season.is_season_start_too_late() {
+				return Err(Error::<T>::SeasonStartsTooLate.into())
+			}
+
+			if Seasons::<T>::get(season_id).is_none() {
+				return Err(Error::<T>::UnknownSeason.into())
+			}
+
+			let mut maybe_previous_season: Option<
+				Season<<T as frame_system::Config>::BlockNumber>,
+			> = None;
+
+			if season_id > 0 {
+				maybe_previous_season = Seasons::<T>::get(season_id - 1);
+			}
+
+			let maybe_next_season = Seasons::<T>::get(season_id + 1);
+
+			enum UpdateError {
+				OverlappedWithPreviousSeason,
+				OverlappedWithNextSeason,
+				NotFound,
+			}
+
+			let mutate_result = Seasons::<T>::try_mutate(season_id, |maybe_season| {
+				if let Some(existing_season) = maybe_season {
+					if let Some(previous_season) = maybe_previous_season {
+						if Season::are_seasons_overlapped(&previous_season, &season) {
+							return Err(UpdateError::OverlappedWithPreviousSeason)
+						}
+					}
+
+					if let Some(next_season) = maybe_next_season {
+						if Season::are_seasons_overlapped(&season, &next_season) {
+							return Err(UpdateError::OverlappedWithNextSeason)
+						}
+					}
+
+					existing_season.end = season.end;
+					existing_season.start = season.start;
+					existing_season.early_access_start = season.early_access_start;
+					existing_season.max_mints = season.max_mints;
+					existing_season.max_mythical_mints = season.max_mythical_mints;
+					Ok(())
+				} else {
+					Err(UpdateError::NotFound)
+				}
+			});
+
+			match mutate_result {
+				Err(UpdateError::OverlappedWithPreviousSeason) =>
+					return Err(Error::<T>::EarlyAccessStartsTooEarly.into()),
+				Err(UpdateError::OverlappedWithNextSeason) =>
+					return Err(Error::<T>::SeasonEndsTooLate.into()),
+				Err(UpdateError::NotFound) => return Err(Error::<T>::UnknownSeason.into()),
+				Ok(_) => {},
+			}
+
+			Self::deposit_event(Event::SeasonUpdated(season, season_id));
+
+			Ok(())
```

Comment:
Let's reduce complexity with:
```suggestion
			Seasons::<T>::try_mutate(season_id, |maybe_season| {
				if let Some(prev_season) = Self::seasons(season_id - 1) {
					ensure!(prev_season.end < season.early_access_start, Error::<T>::EarlyAccessStartsTooEarly);
				}
				if let Some(next_season) = Self::seasons(season_id + 1) {
					ensure!(season.end < next_season.early_access_start, Error::<T>::SeasonEndsTooLate);
				}

				let existing_season = maybe_season.as_mut().ok_or(Error::<T>::UnknownSeason)?;
				*existing_season = season.clone();
				Self::deposit_event(Event::SeasonUpdated(season, season_id));
				Ok(())
			})
```

Reply:
This is much better. Thanks!